### PR TITLE
updates to maya wrapper objects due to MayaPolicy design

### PR DESCRIPTION
**Pull request by AmitKumarDas**
### 1. Why is this change necessary ?

- Maya wrappers for K8s objects need to be updated
to include the design due to following:
  - https://github.com/openebs/openebs/issues/1010

fixes openebs/openebs#1012

### 2. How does this change address the issue ?

- Updates couple of maya structures
- Adds new maya structures
- Adds new unit tests
- Refactors unit tests
- Changes are done keeping **Object Oriented Programming** (_OOP_) in mind

### 3. How to verify this change ?

- run `go test -v` at pkg/maya
- alternatively make test will work fine
- unit tests can be read to understand better

### 4. What side effects does this change have ?

- none

### 5. Unit Test results

```bash
vagrant@maya-01:/opt/gopath/src/github.com/openebs/maya/pkg/maya$ go test -v
=== RUN   TestMayaBytes
=== RUN   TestMayaBytes/blank_yaml
=== RUN   TestMayaBytes/hello_yaml
--- PASS: TestMayaBytes (0.00s)
    --- PASS: TestMayaBytes/blank_yaml (0.00s)
    --- PASS: TestMayaBytes/hello_yaml (0.00s)
=== RUN   TestMayaAnyK8sGenerateService
=== RUN   TestMayaAnyK8sGenerateService/blank_service
=== RUN   TestMayaAnyK8sGenerateService/hello_service
=== RUN   TestMayaAnyK8sGenerateService/invalid_service
=== RUN   TestMayaAnyK8sGenerateService/valid_service
--- PASS: TestMayaAnyK8sGenerateService (0.00s)
    --- PASS: TestMayaAnyK8sGenerateService/blank_service (0.00s)
    --- PASS: TestMayaAnyK8sGenerateService/hello_service (0.00s)
    --- PASS: TestMayaAnyK8sGenerateService/invalid_service (0.00s)
    --- PASS: TestMayaAnyK8sGenerateService/valid_service (0.00s)
=== RUN   TestMayaAnyK8sGenerateServiceTemplated
=== RUN   TestMayaAnyK8sGenerateServiceTemplated/templated_service
--- PASS: TestMayaAnyK8sGenerateServiceTemplated (0.00s)
    --- PASS: TestMayaAnyK8sGenerateServiceTemplated/templated_service (0.00s)
=== RUN   TestMayaAnyK8sGenerateDeployment
=== RUN   TestMayaAnyK8sGenerateDeployment/valid_deployment
=== RUN   TestMayaAnyK8sGenerateDeployment/blank_deployment
=== RUN   TestMayaAnyK8sGenerateDeployment/hello_deployment
=== RUN   TestMayaAnyK8sGenerateDeployment/invalid_deployment
--- PASS: TestMayaAnyK8sGenerateDeployment (0.00s)
    --- PASS: TestMayaAnyK8sGenerateDeployment/valid_deployment (0.00s)
    --- PASS: TestMayaAnyK8sGenerateDeployment/blank_deployment (0.00s)
    --- PASS: TestMayaAnyK8sGenerateDeployment/hello_deployment (0.00s)
    --- PASS: TestMayaAnyK8sGenerateDeployment/invalid_deployment (0.00s)
=== RUN   TestMayaAnyK8sGenerateDeploymentTemplated
=== RUN   TestMayaAnyK8sGenerateDeploymentTemplated/templated_deployment
--- PASS: TestMayaAnyK8sGenerateDeploymentTemplated (0.00s)
    --- PASS: TestMayaAnyK8sGenerateDeploymentTemplated/templated_deployment (0.00s)
=== RUN   TestMayaAnyK8sGenerateCofigMap
=== RUN   TestMayaAnyK8sGenerateCofigMap/blank_yaml_is_invalid_configmap
=== RUN   TestMayaAnyK8sGenerateCofigMap/hello_yaml_is_invalid_configmap
=== RUN   TestMayaAnyK8sGenerateCofigMap/valid_configmap
--- PASS: TestMayaAnyK8sGenerateCofigMap (0.01s)
    --- PASS: TestMayaAnyK8sGenerateCofigMap/blank_yaml_is_invalid_configmap (0.00s)
    --- PASS: TestMayaAnyK8sGenerateCofigMap/hello_yaml_is_invalid_configmap (0.00s)
    --- PASS: TestMayaAnyK8sGenerateCofigMap/valid_configmap (0.00s)
=== RUN   TestMayaAnyK8sGenerateCofigMapTemplated
=== RUN   TestMayaAnyK8sGenerateCofigMapTemplated/a_templated_as_well_as_valid_configmap
--- PASS: TestMayaAnyK8sGenerateCofigMapTemplated (0.00s)
    --- PASS: TestMayaAnyK8sGenerateCofigMapTemplated/a_templated_as_well_as_valid_configmap (0.00s)
=== RUN   TestMayaConfigMapLoad
=== RUN   TestMayaConfigMapLoad/hello_config_map
=== RUN   TestMayaConfigMapLoad/valid_config_map
=== RUN   TestMayaConfigMapLoad/blank_config_map
--- PASS: TestMayaConfigMapLoad (0.00s)
    --- PASS: TestMayaConfigMapLoad/hello_config_map (0.00s)
    --- PASS: TestMayaConfigMapLoad/valid_config_map (0.00s)
    --- PASS: TestMayaConfigMapLoad/blank_config_map (0.00s)
=== RUN   TestMayaConfigMapLoadEmbeddedK8s
=== RUN   TestMayaConfigMapLoadEmbeddedK8s/config_map_without_embedded_k8s
=== RUN   TestMayaConfigMapLoadEmbeddedK8s/config_map_with_embedded_k8s
--- PASS: TestMayaConfigMapLoadEmbeddedK8s (0.00s)
    --- PASS: TestMayaConfigMapLoadEmbeddedK8s/config_map_without_embedded_k8s (0.00s)
    --- PASS: TestMayaConfigMapLoadEmbeddedK8s/config_map_with_embedded_k8s (0.00s)
=== RUN   TestMayaServiceLoad
=== RUN   TestMayaServiceLoad/blank_service
=== RUN   TestMayaServiceLoad/hello_service
=== RUN   TestMayaServiceLoad/valid_service
--- PASS: TestMayaServiceLoad (0.00s)
    --- PASS: TestMayaServiceLoad/blank_service (0.00s)
    --- PASS: TestMayaServiceLoad/hello_service (0.00s)
    --- PASS: TestMayaServiceLoad/valid_service (0.00s)
=== RUN   TestMayaContainerLoad
--- PASS: TestMayaContainerLoad (0.00s)
=== RUN   TestMayaDeploymentLoad
=== RUN   TestMayaDeploymentLoad/valid_yaml
=== RUN   TestMayaDeploymentLoad/blank_yaml
=== RUN   TestMayaDeploymentLoad/hello_yaml
--- PASS: TestMayaDeploymentLoad (0.00s)
    --- PASS: TestMayaDeploymentLoad/valid_yaml (0.00s)
    --- PASS: TestMayaDeploymentLoad/blank_yaml (0.00s)
    --- PASS: TestMayaDeploymentLoad/hello_yaml (0.00s)
=== RUN   TestMayaDeploymentAddContainer
=== RUN   TestMayaDeploymentAddContainer/valid_struct_yaml
=== RUN   TestMayaDeploymentAddContainer/blank_yaml
=== RUN   TestMayaDeploymentAddContainer/hello_yaml
--- PASS: TestMayaDeploymentAddContainer (0.01s)
    --- PASS: TestMayaDeploymentAddContainer/valid_struct_yaml (0.00s)
    --- PASS: TestMayaDeploymentAddContainer/blank_yaml (0.00s)
    --- PASS: TestMayaDeploymentAddContainer/hello_yaml (0.00s)
PASS
ok  	github.com/openebs/maya/pkg/maya	0.121s
```

### CODE REVIEW

**Comment by gkGaneshR** [comment]
@AmitKumarDas  Can you tell the need to have both string and []byte type of yaml as part of struct variable. I understand that byte is required template parsing, but if it can be converted back and forth, a temporary variable would do good, right? Will it not add duplication of data leading to occupy more memory since they are part of same struct var? Is there any other particular reason for that?

**Comment by AmitKumarDas** [comment]
@gkGaneshR I agree there is some element of redundancy. However, it enables in UT.
I might also want the original string to create a hash.
Now we can use this hash to find duplicate policies or templates if any.
However, this hashing stuff is futuristic. As I have learned, it is most important to be able to easily write UT.

File: pkg/maya/maya.go [hunk 1]
```diff
@@ -27,22 +29,74 @@ import (
...
+	// that might have been set in above Yaml property.
+	MayaPlaceholders
+
+	// YmlBytes represents above yaml in byte slice format
+	//
+	// NOTE:
+	//  This is generated by using the templated yaml
+	// (i.e. Yaml property) & placeholders (i.e.
+	// MayaPlaceholders property)
+	YmlBytes []byte
```
Comment (gkGaneshR) [inline pkg/maya/maya.go]: @AmitKumarDas , I couldn't understand how this gets enabled on UT. The variable m* MayaYaml is a global variable under package maya which has redundant piece of variables - Yaml string and YmlBytes []byte. Can u pls explain?

Comment (prateekpandey14) [inline pkg/maya/maya.go]: Can this be replaced by `[ ] string` ?

Comment (prateekpandey14) [inline pkg/maya/maya.go]: Can we replace `[]byte` with `[]string` ?

Comment (prateekpandey14) [reply pkg/maya/maya.go]: Then convert it to [] byte before Unmarshalling to any object type

Comment (AmitKumarDas) [reply pkg/maya/maya.go]: Thanks @prateekpandey14 @gkGaneshR for the reviews. There are 3 properties in MayaYaml struct. Bytes is derived from placeholders & string version of yaml. In some of the cases bytes are derived from string yaml only. However, all these explanation still begs the question on why bytes, as it is redundant.

To answer it, let us think from object point of view. There are cases where the object i.e. MayaYaml needs to be initialized out of bytes directly.  In these cases what should be the struct design?

Let us not think only from a behavior p.o.v. What about object construction & object composition?
Thinking only from behavior p.o.v will lead to procedural code that fails the open-closed principle.

MayaYaml is constructed in 2 ways.
Way 1 - by providing placeholder & string yaml e.g. Load() of MayaConfigMap
Way 2 - by providing bytes[] only e.g. NewMayaServiceByByte()

If you look at the _test.go code, you will find the motive for doing this.

**Pull request merged**